library/ssl_tls.c

```diff
@@ -5818,6 +5863,56 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
 
     mbedtls_platform_zeroize( handshake,
                               sizeof( mbedtls_ssl_handshake_params ) );
+
+#if defined(MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH)
+    /* If the buffers are too big - reallocate */
+    int modified = 0;
+    unsigned char* resized_buffer = NULL;
+    uint32_t buf_len = mbedtls_ssl_get_input_buflen( ssl );
+
+    if( ssl->in_buf != NULL && ssl->in_buf_len > buf_len && ssl->in_left < buf_len )
+    {
+        resized_buffer = mbedtls_calloc( 1,  buf_len );
+        if( resized_buffer != NULL )
+        {
+            memcpy( resized_buffer, ssl->in_buf, buf_len );
+            mbedtls_platform_zeroize( ssl->in_buf, ssl->in_buf_len );
+            mbedtls_free( ssl->in_buf );
+
+            ssl->in_buf = resized_buffer;
+            ssl->in_buf_len = buf_len;
+            modified = 1;
+        }
+        else
+        {
+           MBEDTLS_SSL_DEBUG_MSG( 1, ( "output buffer resizing failed - out of memory" ) );
+        }
+    }
+
+    buf_len = mbedtls_ssl_get_output_buflen( ssl );
+    if( ssl->out_buf != NULL && ssl->out_buf_len > mbedtls_ssl_get_output_buflen( ssl ) && ssl->out_left < buf_len )
+    {
+        resized_buffer = mbedtls_calloc( 1, buf_len );
+        if( resized_buffer != NULL )
+        {
+            memcpy( resized_buffer, ssl->out_buf, buf_len );
+            mbedtls_platform_zeroize( ssl->out_buf, ssl->out_buf_len );
+            mbedtls_free( ssl->out_buf );
```

Comment:
Do we really need to preserve the contents of the buffer here? What is present? Please add a C comment justifying the preservation of the data.

Reply:
Quoting Janos: 

> Mbed TLS processes TLS records one by one and does not have more than a single
> record in the I/O buffers at a given time. When the handshake completes and the
> resizing happens it is safe to assume that both buffers are empty.
> 
> On the other hand, it processes incoming DTLS data datagram by datagram. This
> means that the input buffer needs to be able to hold a whole datagram. A
> datagram is allowed to have several DTLS records in it.
> 
> This has two consequences for DTLS:
> 
> 1. after the handshake it is not safe to assume anymore that the I/O buffers are
>    empty
> [...]
Should I add a comment: `Because of the way mbedTLS processes datagrams and the fact that a datagram is allowed to have several records in it, it is possible that the I/O buffers are not empty at this stage`?